scan-payment/src/main/java/com/getbouncer/scan/payment/analyzer/NameDetectAnalyzer.kt

```diff
@@ -0,0 +1,268 @@
+     * tie-breaking with confidence
+     */
+    private fun processNMSCluster(charClusters: List<AlphabetDetect.Prediction>): Char {
+        var candidateLetter = 0.toChar()
+        var candidateLetterConfidence = 0f
+        var candidateConsecutiveCount = 0
+        var currentConsecutiveCount = 0
+        var currentLetterMaxConfidence = 0f
+        var lastSeenLetter = 0.toChar()
+        for (characterPrediction in charClusters) {
+            if (lastSeenLetter == characterPrediction.character) {
+                currentConsecutiveCount += 1
+                currentLetterMaxConfidence =
+                    Math.max(currentLetterMaxConfidence, characterPrediction.confidence)
+            } else {
+                currentConsecutiveCount = 1
+                currentLetterMaxConfidence = characterPrediction.confidence
+                lastSeenLetter = characterPrediction.character
+            }
+            if (currentConsecutiveCount == candidateConsecutiveCount && currentLetterMaxConfidence > candidateLetterConfidence ||
+                currentConsecutiveCount > candidateConsecutiveCount
+            ) {
+                candidateLetterConfidence = currentLetterMaxConfidence
+                candidateLetter = characterPrediction.character
+                candidateConsecutiveCount = currentConsecutiveCount
+            }
+        }
+        return if (candidateLetterConfidence > CHAR_CONFIDENCE_THRESHOLD) {
+            candidateLetter
+        } else ' '
+    }
+
+    /**
+     * Black magic to calculate the "width" (in terms of prediction index) for each space
+     */
+    private fun getSpacesWidth(spaces: List<Int>): Int {
+        if (spaces.isEmpty()) {
+            return 10
+        }
+        val slice = spaces.subList(1, spaces.size - 1).sorted()
+        val p25Index = slice.size * 25 / 100
+        val p25 = slice[p25Index]
+        val pmax = slice[slice.size - 1]
+        val pmax2 = if (slice.size >= 2) slice[slice.size - 2] else pmax
+        return if (pmax == pmax2 && pmax == p25) {
+            pmax + 1
+        } else if ((pmax - pmax2) * 2 <= pmax2 - p25) {
+            pmax2
+        } else {
+            pmax
+        }
+    }
+
+    /**
+     * Accepts the output from hard NMS and produces the predicted word
+     */
+    private fun processNMSResults(
+        indices: List<Int>,
+        predictions: List<CharPredictionWithBox>
+    ): String? {
+        val word = StringBuilder()
+        val intermediateChars: MutableList<Char> = ArrayList()
+        val charClusters: MutableList<AlphabetDetect.Prediction> = ArrayList()
+        val spaces: MutableList<Int> = ArrayList()
+
+        var currentSpaceClusterSize = 0
+        val debugWord = StringBuilder()
+        for (index in indices) {
+            val prediction: CharPredictionWithBox = predictions[index]
+            debugWord.append(prediction.characterPrediction.character)
```

Comment:
If you call this method with a list of predictions filtered by sorted indices, you can just use the list of predictions.

```suggestion
        predictions.forEach {
            debugWord.append(it.characterPrediction.character)
```